README.md
# thorken
[ ![Codeship Status for RisingStack/thorken](https://codeship.com/projects/f14d16b0-44c7-0133-4946-4686174fbfc9/status?branch=master)](https://codeship.com/projects/104466)  

`Redis` and `JWT` token based session manager.

## Usage

You need [redis](http://redis.io) to use this package.  

`npm install --save thorken`

### Configure

```javascript
var Thorken = require('thorken')
var session = new Thorken({
  jwtSecret: 'secret',
  [namespace]: 'ts',
  [redis]: ioredisInstance,
  [cleanupManual]: false
})

```

### Create

```javascript
session.create({
  uid: '1',
  [ttl]: 7200,
  [ip]: '127.0.0.1'
})
.then(function (jwtToken) { ... })
```

### Get

```javascript
session.get('token')
.then(function (session) { ... })
```

### Extend

The second ttl parameter is optional.

```javascript
session.extend('token', 7200)
.then(function (expiresAt) { ... })
```

### Destroy

```javascript
session.destroy('token')
.then(function (isSuccess) { ... })
```

### Cleanup

For manual cleanup.

To clear only expired sessions

```javascript
session.cleanup().then(function () { ... })
```

To clear every session

```javascript
session.cleanup(true).then(function () { ... })
```

### Get user's sessions

```javascript
session.getByUserId('1').then(function (sessions) { ... })
```

### Destroy user's sessions

```javascript
session.destroyUser('1').then(function (isSuccess) { ... })
```

## Server connectors

### Koa middleware

```javascript
var app = require('koa')()
var Thorken = require('thorken')
var protect = require('thorken/src/koa')
var session = new Thorken({ ... })

app.use(protect(session, {
  [extend]: true
}))
```


CHANGELOG.md
<a name="1.3.0"></a>
# 1.3.0 (2015-10-21)


### docs

* docs(reamde): fix extend docs ([585eee5](https://github.com/RisingStack/thorken/commit/585eee5))

### feat

* feat(getByUserId): implement ([85f1274](https://github.com/RisingStack/thorken/commit/85f1274))



<a name="1.2.1"></a>
## 1.2.1 (2015-10-20)


### chore

* chore(package): update dependencies ([1b2d5aa](https://github.com/RisingStack/thorken/commit/1b2d5aa))

### fix

* fix(destroyUser): remove only existing data ([c0d460f](https://github.com/RisingStack/thorken/commit/c0d460f))



<a name="1.2.0"></a>
# 1.2.0 (2015-09-25)


### feat

* feat(destroy): add destroy user ([d426c66](https://github.com/RisingStack/thorken/commit/d426c66))



<a name="1.1.1"></a>
## 1.1.1 (2015-09-25)


### docs

* docs(readme): add install snippet ([a8b7838](https://github.com/RisingStack/thorken/commit/a8b7838))
* docs(readme): add redis requirements ([3129a64](https://github.com/RisingStack/thorken/commit/3129a64))
* docs(readme): fix cleanup interface ([91ebd14](https://github.com/RisingStack/thorken/commit/91ebd14))

### test

* test(koa): cover Koa middleware with unit and e2e tests ([75426bb](https://github.com/RisingStack/thorken/commit/75426bb))



<a name="1.1.0"></a>
# 1.1.0 (2015-09-25)


### feat

* feat(koa): add Koa middleware ([7d3655b](https://github.com/RisingStack/thorken/commit/7d3655b))



<a name="1.0.0"></a>
# 1.0.0 (2015-09-24)


### chore

* chore(package): add linter ([1e75a79](https://github.com/RisingStack/thorken/commit/1e75a79))
* chore(package): initial commit ([178140d](https://github.com/RisingStack/thorken/commit/178140d))
* chore(project): rename project to thorken :) ([eaa14a5](https://github.com/RisingStack/thorken/commit/eaa14a5))
* chore(readme): use javascript ([2376868](https://github.com/RisingStack/thorken/commit/2376868))

### docs

* docs(readme): add CodeShip badge ([099a682](https://github.com/RisingStack/thorken/commit/099a682))
* docs(readme): add create and cleanup docs ([5f9b23a](https://github.com/RisingStack/thorken/commit/5f9b23a))
* docs(readme): add todo features ([4891dde](https://github.com/RisingStack/thorken/commit/4891dde))

### feat

* feat(cleanup): add cleanup ([729092a](https://github.com/RisingStack/thorken/commit/729092a))
* feat(destroy): add destroy feature ([f56069e](https://github.com/RisingStack/thorken/commit/f56069e))
* feat(expire): expire in redis instead of token ([8fbeb3d](https://github.com/RisingStack/thorken/commit/8fbeb3d))
* feat(extend): add extend feature ([9f6162d](https://github.com/RisingStack/thorken/commit/9f6162d))
* feat(get): get session ([29d29d9](https://github.com/RisingStack/thorken/commit/29d29d9))
* feat(tokenSession): add create ([253affe](https://github.com/RisingStack/thorken/commit/253affe))

### fix

* fix(destroy): use zrem to remove from token list set ([ef4afa1](https://github.com/RisingStack/thorken/commit/ef4afa1))

### test

* test(e2e): add create and cleanup tests ([6e07580](https://github.com/RisingStack/thorken/commit/6e07580))





src/e2e.spec.js
var Redis = require('ioredis')
var expect = require('chai').expect
var jwt = require('jsonwebtoken')

var Thorken = require('../index.js')

describe('e2e', () => {
  var redis = new Redis({
    db: 1
  })

  var thorken = new Thorken({
    redis: redis,
    cleanupManual: true,
    jwtSecret: 'secret'
  })

  before(function *() {
    yield redis.flushall()
  })

  describe('#create', () => {
    it('should create a session', function *() {
      var expireAt = Date.now() + 7200000
      var token = yield thorken.create({
        uid: '1'
      })

      var tokenPayload = yield jwt.verify(token, 'secret')
      var props = yield redis.hgetall(thorken.namespaceKey + 't:' + token)

      expect(tokenPayload).to.have.property('uid', '1')

      expect(props).to.have.property('uid', '1')
      expect(props).to.have.property('exp')
      expect(Number(props.exp)).to.be.at.least(expireAt)
    })

    it('should create a session which expires', function *() {
      var token = yield thorken.create({
        uid: '1'
      })

      var ttl = yield redis.pttl(thorken.namespaceKey + 't:' + token)

      expect(ttl).to.be.above(7198000)
      expect(ttl).to.be.below(7200000)
    })

    afterEach(function *() {
      yield redis.flushall()
    })
  })

  describe('#extend', () => {
    var token

    beforeEach(function *() {
      token = yield thorken.create({
        uid: '1',
        ttl: 2
      })
    })

    it('should extend a session expiration', function *() {
      var expireAt = Date.now() + 7200000

      yield thorken.extend(token)

      // expect
      var tokenKey = thorken.namespaceKey + 't:' + token
      var props = yield redis.hgetall(tokenKey)
      var ttl = yield redis.pttl(tokenKey)

      expect(ttl).to.be.above(7198000)
      expect(ttl).to.be.below(7200000)

      expect(Number(props.exp)).to.be.at.least(expireAt)
    })

    it('should handle if token is invalid', function *() {
      try {
        yield thorken.extend('a.a.b')
      } catch (err) {
        expect(err.message).to.be.equal('invalid token')
        return
      }

      throw new Error('unhandled error')
    })

    afterEach(function *() {
      yield redis.flushall()
    })
  })

  describe('#get', () => {
    var token

    beforeEach(function *() {
      token = yield thorken.create({
        uid: '1',
        ip: '192.168.1.1'
      })
    })

    afterEach(function *() {
      yield thorken.destroy(token)
    })

    it('should return with session', function *() {
      var props = yield thorken.get(token)

      expect(props).to.have.property('uid', '1')
      expect(props).to.have.property('ip', '192.168.1.1')
    })

    it('should reject malformed token', function *() {
      try {
        yield thorken.get('invalid token')
      } catch (err) {
        expect(err.message).to.be.equal('jwt malformed')
        return
      }

      throw new Error('unhandled error')
    })

    it('should reject invalid token', function *() {
      try {
        yield thorken.get('a.a.b')
      } catch (err) {
        expect(err.message).to.be.equal('invalid token')
        return
      }

      throw new Error('unhandled error')
    })

    it('should reject unknown token', function *() {
      var token = jwt.sign({
        uid: '1'
      }, thorken.jwtSecret)

      try {
        yield thorken.get(token)
      } catch (err) {
        expect(err.message).to.be.equal('unknown token')
        return
      }

      throw new Error('unhandled error')
    })

    afterEach(function *() {
      yield redis.flushall()
    })
  })

  describe('#getByUserId', () => {
    var token1
    var token2

    beforeEach(function *() {
      var result = yield {
        token1: thorken.create({
          uid: '1',
          ip: '192.168.1.1'
        }),

        token2: thorken.create({
          uid: '1',
          ip: '192.168.1.2'
        })
      }

      token1 = result.token1
      token2 = result.token2
    })

    afterEach(function *() {
      yield [thorken.destroy(token1), thorken.destroy(token2)]
    })

    it('should return with user\'s sessions', function *() {
      var sessions1 = yield thorken.getByUserId('1')
      var sessions2 = yield thorken.getByUserId('2')

      expect(sessions1.length).to.be.equal(2)
      expect(sessions1[0]).property('uid', '1')
      expect(sessions1[0]).property('ip')
      expect(sessions1[1]).property('uid', '1')
      expect(sessions1[1]).property('ip')

      expect(sessions2.length).to.be.equal(0)
    })
  })

  describe('#destroy', () => {
    var token

    beforeEach(function *() {
      token = yield thorken.create({
        uid: '1'
      })
    })

    it('should return with session', function *() {
      var iSuccess = yield thorken.destroy(token)

      // expect
      var userTokens = yield redis.smembers(thorken.namespaceKey + 'u:1')
      var tokens = yield redis.zrangebyscore(thorken.namespaceKey + 't:list', 0, '+inf')
      var props = yield redis.hgetall(thorken.namespaceKey + 't:' + token)

      expect(iSuccess).to.be.true
      expect(userTokens).to.be.eql([])
      expect(tokens).to.be.eql([])
      expect(props).to.be.eql({})
    })

    afterEach(function *() {
      yield redis.flushall()
    })
  })

  describe('#destroy user', () => {
    var token1
    var token3

    beforeEach(function *() {
      token1 = yield thorken.create({
        uid: '1'
      })

      yield thorken.create({
        uid: '1'
      })

      token3 = yield thorken.create({
        uid: '2'
      })
    })

    it('should remove user and user\'s tokens', function *() {
      var iSuccess = yield thorken.destroyUser('1')

      // expect
      var userTokens = yield redis.smembers(thorken.namespaceKey + 'u:1')
      var tokens = yield redis.zrangebyscore(thorken.namespaceKey + 't:list', 0, '+inf')
      var props = yield redis.hgetall(thorken.namespaceKey + 't:' + token1)

      expect(iSuccess).to.be.true
      expect(userTokens).to.be.eql([])
      expect(tokens).to.be.eql([
        '2:' + token3
      ])
      expect(props).to.be.eql({})
    })

    afterEach(function *() {
      yield redis.flushall()
    })
  })

  describe('#cleanup', () => {
    beforeEach(function *() {
      yield thorken.create({
        uid: '1',
        ttl: 0
      })
    })

    it('should remove all sessions', function *() {
      yield thorken.cleanup(true)

      var key = yield redis.randomkey()

      expect(key).to.be.null
    })

    it('should remove expired sessions', function *() {
      yield thorken.cleanup()

      var key = yield redis.randomkey()

      expect(key).to.be.null
    })

    it('should remove only expired sessions', function *() {
      var token2 = yield thorken.create({
        uid: '1'
      })

      yield thorken.cleanup()

      var tokens = yield redis.smembers(thorken.namespaceKey + 'u:1')

      expect(tokens).to.be.eql([
        token2
      ])
    })
  })
})


src/koa/koa.spec.js
var expect = require('chai').expect

var middleware = require('./koa')
var Thorken = require('../../index')

describe('koa middleware', () => {
  var thorken
  var next
  var ctx

  beforeEach(function () {
    thorken = new Thorken({
      jwtSecret: 'secret'
    })

    // koa mocks
    next = function *() {}

    ctx = {
      headers: {
        authorization: 'Bearer token'
      },
      state: {
        user: {}
      }
    }
  })

  it('should validate and extend token and set state', function *() {
    var getStub = this.sandbox.stub(thorken, 'get').returnsWithResolve({
      uid: '1'
    })
    var extendStub = this.sandbox.stub(thorken, 'extend').returnsWithResolve()
    var mw = middleware(thorken)

    yield mw.call(ctx, next)

    expect(getStub).to.be.calledWith('token')
    expect(extendStub).to.be.calledWith('token')

    expect(ctx.state).to.be.eql({
      user: {
        id: '1'
      }
    })
  })

  it('should not extend token', function *() {
    this.sandbox.stub(thorken, 'get').returnsWithResolve({
      uid: '1'
    })
    var extendStub = this.sandbox.stub(thorken, 'extend').returnsWithResolve()

    var mw = middleware(thorken, {
      extend: false
    })

    yield mw.call(ctx, next)

    expect(extendStub).to.be.not.called
  })

  it('should return with unauthorized', function *() {
    this.sandbox.stub(thorken, 'get').returnsWithReject(new Error('invalid token'))
    var extendStub = this.sandbox.stub(thorken, 'extend').returnsWithResolve()

    var mw = middleware(thorken)

    try {
      yield mw.call(ctx, next)
    } catch (err) {
      expect(extendStub).to.be.not.called
      expect(err.message).to.be.eql('invalid token')
      return
    }

    throw new Error('uncatched error')
  })
})


src/testSetup.spec.js
'use strict'

var sinon = require('sinon')
var chai = require('chai')
var sinonChai = require('sinon-chai')
var debug = require('debug')

debug.enable('thorken:*')

before(function () {
  chai.use(sinonChai)

  sinon.stub.returnsWithResolve = function (data) {
    return this.returns(Promise.resolve(data))
  }

  sinon.stub.returnsWithReject = function (error) {
    return this.returns(Promise.reject(error))
  }
})

beforeEach(function () {
  this.sandbox = sinon.sandbox.create()
})

afterEach(function () {
  this.sandbox.restore()
})


src/koa/koa.js
var _ = require('lodash')
var Thorken = require('../thorken')

/**
* @method factory
* @param {Thorken} thorken
* @param {Object} opts {
*   [extend]: true
* }
* @return {Generator} middleware
*/
function factory (thorken, opts) {
  var options = _.defaults(opts || {}, {
    extend: true
  })

  if (!(thorken instanceof Thorken)) {
    throw new Error('instance of Thorken is required')
  }

  // middleware
  return function *(next) {
    var token = (this.headers.authorization || '').substring(7)
    var session

    // get session
    try {
      session = yield thorken.get(token)
    } catch (err) {
      err.status = 401
      throw err
    }

    // extend token's expiration
    if (options.extend) {
      yield thorken.extend(token)
    }

    this.state.user = _.merge(this.state.user, {
      id: session.uid
    })

    yield next
  }
}

module.exports = factory


src/tokenSession.js
var _ = require('lodash')
var Redis = require('ioredis')
var jwt = require('jsonwebtoken')
var Promise = require('promise')

var PREFIX = {
  NAMESPACE: 'n:',
  USER: 'u:',
  TOKEN: 't:'
}

var DEFAULT = {
  NAMESPACE: 'ts',
  TTL: 7200,
  CLEANUP_INTERVAL: 300000 // 5 minutes
}

/**
* @class TokenSession
* @param {Object} opts {
*  [namespace]: String,
*  [redis]: Redis, ioredis instance,
*  jwtSecret: String,
*  [cleanupManual]: Boolean, default: false
* }
*/
function TokenSession (opts) {
  var _this = this

  opts = opts || {}
  opts = _.defaults(opts, {
    namespace: DEFAULT.NAMESPACE,
    cleanupInterval: DEFAULT.CLEANUP_INTERVAL,
    cleanupManual: false
  })

  if (!opts.jwtSecret) {
    throw new Error('jwtSecret is required')
  }

  _this.redis = opts.redis || new Redis()
  _this.namespace = opts.namespace
  _this.namespaceKey = PREFIX.NAMESPACE + _this.namespace + ':'
  _this.jwtSecret = opts.jwtSecret

  if (!opts.cleanupManual) {
    _this.cleanupIntervalId = setInterval(function () {
      _this.cleanup()
    }, opts.cleanupInterval)
  }
}

/**
* @method create
* @param {Object} opts {
*   uid: String
*   [ttl]: Number in second
*   [ip]: String
* }
* @return {String} token, jwt
*/
TokenSession.prototype.create = function (opts) {
  var _this = this

  opts = opts || {}
  opts = _.defaults(opts, {
    ttl: DEFAULT.TTL
  })

  if (!opts.uid) {
    return Promise.reject(new Error('uid is required'))
  }

  if (!_.isNumber(opts.ttl)) {
    return Promise.reject(new Error('ttl is required and should be a Number'))
  }

  opts.uid = String(opts.uid)

  // create JWT token
  var token = jwt.sign({
    uid: opts.uid,
    ts: Date.now()
  }, _this.jwtSecret)

  var userKey = _this.namespaceKey + PREFIX.USER + opts.uid
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'
  var tokenListValue = opts.uid + ':' + token
  var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token

  var expiresAt = Date.now() + (opts.ttl * 1000)
  var tokenPayload = {
    uid: opts.uid,
    exp: expiresAt
  }

  if (opts.ip) {
    tokenPayload.ip = opts.ip
  }

  return _this.redis
    .multi()

    // add token to the list by score
    // order by score makes easy to get and remove expired ones
    .zadd(tokenListKey, expiresAt, tokenListValue)

    // add token to user
    .sadd(userKey, token)

    // store token props
    .hmset(tokenKey, tokenPayload)
    .exec()
    .then(function () {
      return _this.redis.pexpireat(tokenKey, expiresAt)
        .then(function (result) {
          if (result !== 1) {
            throw new Error('cannot set expiration on token')
          }

          return token
        })
    })
}

/**
* Get token properties
* @method get
*/
TokenSession.prototype.get = function (token) {
  var _this = this
  var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token

  return Promise.all([
    TokenSession.jwtVerify(token, _this.jwtSecret),
    _this.redis.hgetall(tokenKey)
  ])
    .then(function (results) {
      var props = results[1]

      if (_.isEqual(props, {})) {
        throw new Error('unknown token')
      }

      props.exp = Number(props.exp)

      return results[1]
    })
}

/**
* Extend token's expiration
* @method extend
*/
TokenSession.prototype.extend = function (token, ttl) {
  ttl = _.isNumber(ttl) ? ttl : DEFAULT.TTL

  var _this = this

  return _this.get(token)
    .then(function (props) {
      var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token
      var expiresAt = Date.now() + (ttl * 1000)

      props.exp = expiresAt

      return _this.redis
        .multi()
        .hmset(tokenKey, props)
        .pexpireat(tokenKey, expiresAt)
        .exec()
        .then(function () {
          return expiresAt
        })
    })
}

/**
* Destory token
* @method destroy
*/
TokenSession.prototype.destroy = function (token) {
  var _this = this
  var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token
  var tokenPayload = jwt.verify(token, _this.jwtSecret)
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'
  var userKey = _this.namespaceKey + PREFIX.USER + tokenPayload.uid
  var tokenListValue = tokenPayload.uid + ':' + token

  return this.redis
    .multi()
    .del(tokenKey)
    .srem(userKey, token)
    .zrem(tokenListKey, tokenListValue)
    .exec()
    .then(function (results) {
      return results[0][1] === 1
    })
}

/**
* Remove expired tokens
* @method cleanup
*/
TokenSession.prototype.cleanup = function (cleanupAll) {
  var _this = this
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'
  var to = Date.now()

  if (cleanupAll) {
    to = '+inf'
  }

  return this.redis

    // get expired tokens
    .zrangebyscore(tokenListKey, 0, to)
      .then(function (expiredItems) {
        // extract token keys from expired items
        var tokenKeys = expiredItems.map(function (item) {
          var tmp = item.split(':')
          var token = tmp[1]
          var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token

          return tokenKey
        })

        // remove from token list and token properties
        var multi = _this.redis.multi()
          .zremrangebyscore(tokenListKey, 0, to)

        if (cleanupAll) {
          multi = multi.del(tokenKeys)
        }

        // remove tokens from users
        expiredItems.forEach(function (item) {
          var tmp = item.split(':')
          var uid = tmp[0]
          var token = tmp[1]
          var userKey = _this.namespaceKey + PREFIX.USER + uid

          multi = multi.srem(userKey, token)
        })

        // return with number of removed tokens
        return multi
          .exec()
          .then(function (results) {
            return results[0]
          })
      })
}

/**
* @method jwtVerify
* @param {String} token
*/
TokenSession.jwtVerify = function () {
  var args = Array.prototype.slice.call(arguments)

  return new Promise(function (resolve, reject) {
    args.push(function (err, payload) {
      if (err) {
        return reject(err)
      }

      resolve(payload)
    })

    jwt.verify.apply(jwt, args)
  })
}

module.exports = TokenSession


src/thorken.js
var _ = require('lodash')
var Redis = require('ioredis')
var jwt = require('jsonwebtoken')
var Promise = require('promise')

var PREFIX = {
  NAMESPACE: 'n:',
  USER: 'u:',
  TOKEN: 't:'
}

var DEFAULT = {
  NAMESPACE: 'ts',
  TTL: 7200,
  CLEANUP_INTERVAL: 300000 // 5 minutes
}

/**
* @class Thorken
* @param {Object} opts {
*  [namespace]: String,
*  [redis]: Redis, ioredis instance,
*  jwtSecret: String,
*  [cleanupManual]: Boolean, default: false
* }
*/
function Thorken (opts) {
  var _this = this

  opts = opts || {}
  opts = _.defaults(opts, {
    namespace: DEFAULT.NAMESPACE,
    cleanupInterval: DEFAULT.CLEANUP_INTERVAL,
    cleanupManual: false
  })

  if (!opts.jwtSecret) {
    throw new Error('jwtSecret is required')
  }

  _this.redis = opts.redis || new Redis()
  _this.namespace = opts.namespace
  _this.namespaceKey = PREFIX.NAMESPACE + _this.namespace + ':'
  _this.jwtSecret = opts.jwtSecret

  if (!opts.cleanupManual) {
    _this.cleanupIntervalId = setInterval(function () {
      _this.cleanup()
    }, opts.cleanupInterval)
  }
}

/**
* @method create
* @param {Object} opts {
*   uid: String
*   [ttl]: Number in second
*   [ip]: String
* }
* @return {String} token, jwt
*/
Thorken.prototype.create = function (opts) {
  var _this = this

  opts = opts || {}
  opts = _.defaults(opts, {
    ttl: DEFAULT.TTL
  })

  if (!opts.uid) {
    return Promise.reject(new Error('uid is required'))
  }

  if (!_.isNumber(opts.ttl)) {
    return Promise.reject(new Error('ttl is required and should be a Number'))
  }

  opts.uid = String(opts.uid)

  // create JWT token
  var token = jwt.sign({
    uid: opts.uid,
    ts: Date.now()
  }, _this.jwtSecret)

  var userKey = _this.namespaceKey + PREFIX.USER + opts.uid
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'
  var tokenListValue = opts.uid + ':' + token
  var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token

  var expiresAt = Date.now() + (opts.ttl * 1000)
  var tokenPayload = {
    uid: opts.uid,
    exp: expiresAt
  }

  if (opts.ip) {
    tokenPayload.ip = opts.ip
  }

  return _this.redis
    .multi()

    // add token to the list by score
    // order by score makes easy to get and remove expired ones
    .zadd(tokenListKey, expiresAt, tokenListValue)

    // add token to user
    .sadd(userKey, token)

    // store token props
    .hmset(tokenKey, tokenPayload)
    .exec()
    .then(function () {
      return _this.redis.pexpireat(tokenKey, expiresAt)
        .then(function (result) {
          if (result !== 1) {
            throw new Error('cannot set expiration on token')
          }

          return token
        })
    })
}

/**
* Get token properties
* @method get
*/
Thorken.prototype.get = function (token) {
  var _this = this
  var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token

  return Promise.all([
    Thorken.jwtVerify(token, _this.jwtSecret),
    _this.redis.hgetall(tokenKey)
  ])
    .then(function (results) {
      var props = results[1]

      if (_.isEqual(props, {})) {
        throw new Error('unknown token')
      }

      props.exp = Number(props.exp)

      return results[1]
    })
}

/**
* Get sessions by user id
* @method getByUserId
*/
Thorken.prototype.getByUserId = function (userId) {
  var _this = this
  var userKey = _this.namespaceKey + PREFIX.USER + userId

  return _this.redis.smembers(userKey)
    .then(function (tokens) {
      var hgetTokens = tokens.map(function (token) {
        var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token
        return _this.redis.hgetall(tokenKey)
      })

      return Promise.all(hgetTokens)
    })
}

/**
* Extend token's expiration
* @method extend
*/
Thorken.prototype.extend = function (token, ttl) {
  ttl = _.isNumber(ttl) ? ttl : DEFAULT.TTL

  var _this = this

  return _this.get(token)
    .then(function (props) {
      var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token
      var expiresAt = Date.now() + (ttl * 1000)

      props.exp = expiresAt

      return _this.redis
        .multi()
        .hmset(tokenKey, props)
        .pexpireat(tokenKey, expiresAt)
        .exec()
        .then(function () {
          return expiresAt
        })
    })
}

/**
* Destory token
* @method destroy
*/
Thorken.prototype.destroy = function (token) {
  var _this = this
  var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token
  var tokenPayload = jwt.verify(token, _this.jwtSecret)
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'
  var userKey = _this.namespaceKey + PREFIX.USER + tokenPayload.uid
  var tokenListValue = tokenPayload.uid + ':' + token

  return this.redis
    .multi()
    .del(tokenKey)
    .srem(userKey, token)
    .zrem(tokenListKey, tokenListValue)
    .exec()
    .then(function (results) {
      return results[0][1] === 1
    })
}

/**
* Destory token
* @method destroyUser
*/
Thorken.prototype.destroyUser = function (uid) {
  var _this = this
  var userKey = _this.namespaceKey + PREFIX.USER + uid
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'

  return _this.redis.smembers(userKey)
    .then(function (tokens) {
      var keysForDel = tokens.map(function (token) {
        return _this.namespaceKey + PREFIX.TOKEN + token
      })

      keysForDel.push(userKey)

      var userTokens = tokens.map(function (token) {
        return uid + ':' + token
      })

      var multi = _this.redis.multi()

      // remove tokens
      if (keysForDel.length) {
        multi = multi.del(keysForDel)
      }

      // remove user's tokens from token list
      if (userTokens.length) {
        multi = multi.zrem(tokenListKey, userTokens)
      }

      return multi
        .exec()
        .then(function (results) {
          return results.every(function (result) {
            return result[0] === null
          })
        })
    })
}

/**
* Remove expired tokens
* @method cleanup
*/
Thorken.prototype.cleanup = function (cleanupAll) {
  var _this = this
  var tokenListKey = _this.namespaceKey + PREFIX.TOKEN + 'list'
  var to = Date.now()

  if (cleanupAll) {
    to = '+inf'
  }

  return this.redis

    // get expired tokens
    .zrangebyscore(tokenListKey, 0, to)
      .then(function (expiredItems) {
        // extract token keys from expired items
        var tokenKeys = expiredItems.map(function (item) {
          var tmp = item.split(':')
          var token = tmp[1]
          var tokenKey = _this.namespaceKey + PREFIX.TOKEN + token

          return tokenKey
        })

        // remove from token list and token properties
        var multi = _this.redis.multi()
          .zremrangebyscore(tokenListKey, 0, to)

        if (cleanupAll) {
          multi = multi.del(tokenKeys)
        }

        // remove tokens from users
        expiredItems.forEach(function (item) {
          var tmp = item.split(':')
          var uid = tmp[0]
          var token = tmp[1]
          var userKey = _this.namespaceKey + PREFIX.USER + uid

          multi = multi.srem(userKey, token)
        })

        // return with number of removed tokens
        return multi
          .exec()
          .then(function (results) {
            return results[0]
          })
      })
}

/**
* @method jwtVerify
* @param {String} token
*/
Thorken.jwtVerify = function () {
  var args = Array.prototype.slice.call(arguments)

  return new Promise(function (resolve, reject) {
    args.push(function (err, payload) {
      if (err) {
        return reject(err)
      }

      resolve(payload)
    })

    jwt.verify.apply(jwt, args)
  })
}

module.exports = Thorken
